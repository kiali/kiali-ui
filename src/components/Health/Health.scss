svg:not(:root).icon-failure,
svg:not(:root).icon-degraded,
svg:not(:root).icon-healthy,
svg:not(:root).icon-na {
  overflow: visible;
}
.tippy-popper[x-placement^='left'] .pf-tooltip-theme .tippy-arrow {
  background-color: #fff;
  border-color: black;
  border-width: 1px;
  border-style: solid;
}
.health_indicator .pf-c-tooltip__content {
  background-color: #fff;
  border-style: solid;
  border-color: black;
  border-width: 1px;
  text-align: left;
}
