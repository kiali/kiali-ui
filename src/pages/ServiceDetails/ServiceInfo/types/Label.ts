export default class Label {
  key: string;
  value: string;

  constructor(key: string, value: string) {
    this.key = key;
    this.value = value;
  }

  isValid() {
    return typeof this.key === 'string' && typeof this.value === 'string';
  }
}
