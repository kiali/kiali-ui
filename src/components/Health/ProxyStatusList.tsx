import * as React from 'react';
import { ProxyStatus } from '../../types/Health';
import { Flex, FlexItem, Split, SplitItem, Text, TextVariants } from '@patternfly/react-core';
import { style } from 'typestyle';

type Props = {
  workloadName: string;
  statuses: ProxyStatus[];
};

const smallStyle = style({ fontSize: '70%' });
const workloadNameStyle = style({ minWidth: '6rem', overflowWrap: 'break-word' });

const proxyStatusSort = (a: ProxyStatus, b: ProxyStatus): number => {
  return a.component > b.component ? 1 : -1;
};

class ProxyStatusList extends React.Component<Props> {
  sortedStatus = () => {
    return this.props.statuses.sort(proxyStatusSort);
  };

  statusList = () => {
    const statusItems = this.sortedStatus().map((value: ProxyStatus, i: number) => {
      return (
        <FlexItem key={'proxy-status-' + i}>
          <Text component={TextVariants.small} className={smallStyle}>
            {value.component + ': ' + value.status}
          </Text>
        </FlexItem>
      );
    });

    return <Flex style={{ flex: 'inlineFlex' }}> {statusItems} </Flex>;
  };

  render() {
    if (this.props.statuses.length > 0) {
      return (
        <Split gutter={'sm'}>
          <SplitItem className={workloadNameStyle}>{this.props.workloadName}</SplitItem>
          <SplitItem>{this.statusList()}</SplitItem>
        </Split>
      );
    } else {
      return [];
    }
  }
}

export default ProxyStatusList;
