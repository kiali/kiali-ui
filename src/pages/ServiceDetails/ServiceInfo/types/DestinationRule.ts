import { DestinationRuleSpec, K8sMetadata } from '../../../../types/IstioObjects';
import Subset from './Subset';

export default class DestinationRule {
  metadata: K8sMetadata;
  spec: DestinationRuleSpec;
  unformattedField: string;

  constructor(metadata: K8sMetadata, spec: DestinationRuleSpec) {
    this.metadata = metadata;
    this.spec = spec;
  }

  isValid() {
    return this.hasValidName() && this.hasSpecData() && this.hasValidSubsets() && this.hasValidHost();
  }

  hasValidName() {
    if (typeof this.metadata.name !== 'string') {
      this.unformattedField = 'Name';
      return false;
    }

    return true;
  }

  hasSpecData() {
    return this.spec !== null;
  }

  hasValidSubsets() {
    if (!this.spec.subsets) {
      return true;
    }

    let valid = this.spec.subsets instanceof Array;
    valid =
      valid &&
      this.spec.subsets.every((subset, i, ary) =>
        new Subset(subset.name, subset.labels, subset.trafficPolicy).isValid()
      );

    if (!valid) {
      this.unformattedField = 'Subsets';
    }

    return valid;
  }

  hasValidHost() {
    if (typeof this.spec.host !== 'string') {
      this.unformattedField = 'Host';
      return false;
    }

    return true;
  }

  formatValidation() {
    if (!this.isValid()) {
      return {
        message: 'This destination rule has format problems in field ' + this.unformattedField,
        severity: 'error',
        path: ''
      };
    }

    return null;
  }
}
