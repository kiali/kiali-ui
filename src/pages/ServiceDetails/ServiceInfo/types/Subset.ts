import { TrafficPolicy } from '../../../../types/IstioObjects';
import Label from './Label';

export default class Subset {
  name: string;
  labels: { [key: string]: string };
  trafficPolicy: TrafficPolicy;

  constructor(name: string, labels: { [key: string]: string }, trafficPolicy: TrafficPolicy) {
    this.name = name;
    this.labels = labels;
    this.trafficPolicy = trafficPolicy;
  }

  public isValid() {
    return this.hasValidName() && this.hasValidLabels() && this.hasValidTrafficPolicy();
  }

  hasValidName() {
    return typeof this.name === 'string';
  }

  hasValidLabels() {
    const valid = Object.keys(this.labels).every((k, i) => new Label(k, this.labels[k]).isValid());
    return this.labels instanceof Object && valid;
  }

  hasValidTrafficPolicy() {
    return true;
  }
}
