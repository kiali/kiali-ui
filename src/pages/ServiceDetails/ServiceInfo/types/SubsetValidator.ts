import { Subset } from '../../../../types/IstioObjects';

export default class SubsetValidator {
  subset: Subset;

  constructor(subset: Subset) {
    this.subset = subset;
  }

  public isValid() {
    return this.hasValidName() && this.hasValidLabels();
  }

  hasValidName() {
    return this.hasStringType(this.subset.name);
  }

  hasValidLabels() {
    const valid = Object.keys(this.subset.labels).every((k, _i) => this.hasValidLabel(k, this.subset.labels[k]));
    return this.subset.labels instanceof Object && valid;
  }

  hasValidLabel(name: string, value: string): boolean {
    return this.hasStringType(name) && this.hasStringType(value);
  }

  hasStringType(value: any) {
    return typeof value === 'string';
  }
}
