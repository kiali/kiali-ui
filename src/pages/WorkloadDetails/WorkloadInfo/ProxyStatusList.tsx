import * as React from 'react';
import { isProxyStatusComponentSynced, isProxyStatusSynced, ProxyStatus } from '../../../types/Health';
import { Stack, StackItem } from '@patternfly/react-core';
import { style } from 'typestyle';
import { PfColors } from '../../../components/Pf/PfColors';

type Props = {
  status?: ProxyStatus;
};

const smallStyle = style({ fontSize: '70%', color: PfColors.White });
const colorStyle = style({ fontSize: '1.1rem', color: PfColors.White });

class ProxyStatusList extends React.Component<Props> {
  statusList = () => {
    if (!this.props.status) {
      return [];
    }

    return [
      { c: 'CDS', s: this.props.status.CDS },
      { c: 'EDS', s: this.props.status.EDS },
      { c: 'LDS', s: this.props.status.LDS },
      { c: 'RDS', s: this.props.status.RDS }
    ].map((value: { c: string; s: string }, i: number) => {
      if (!isProxyStatusComponentSynced(value.s)) {
        const status = value.s ? value.s : '-';
        return (
          <StackItem key={'proxy-status-' + i} className={smallStyle}>
            {value.c + ': ' + status}
          </StackItem>
        );
      } else {
        return null;
      }
    });
  };

  render() {
    if (this.props.status && !isProxyStatusSynced(this.props.status)) {
      return (
        <Stack>
          <StackItem className={colorStyle}>Istio Proxy Status</StackItem>
          {this.statusList()}
        </Stack>
      );
    } else {
      return null;
    }
  }
}

export default ProxyStatusList;
