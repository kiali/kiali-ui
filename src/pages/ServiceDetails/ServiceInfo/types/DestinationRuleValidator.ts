import { DestinationRule } from '../../../../types/IstioObjects';
import SubsetValidator from './SubsetValidator';

export default class DestinationRuleValidator {
  destinationRule: DestinationRule;
  unformattedField: string;

  constructor(destinationRule: DestinationRule) {
    this.destinationRule = destinationRule;
    this.unformattedField = 'none';
  }

  isValid() {
    return this.hasValidName() && this.hasSpecData() && this.hasValidSubsets() && this.hasValidHost();
  }

  hasValidName() {
    if (typeof this.destinationRule.metadata.name !== 'string') {
      this.unformattedField = 'Name';
      return false;
    }

    return true;
  }

  hasSpecData() {
    return this.destinationRule.spec !== null;
  }

  hasValidSubsets() {
    if (!this.destinationRule.spec.subsets) {
      return true;
    }

    let valid = this.destinationRule.spec.subsets instanceof Array;
    valid =
      valid && this.destinationRule.spec.subsets.every((subset, _i, _ary) => new SubsetValidator(subset).isValid());

    if (!valid) {
      this.unformattedField = 'Subsets';
    }

    return valid;
  }

  hasValidHost() {
    if (typeof this.destinationRule.spec.host !== 'string') {
      this.unformattedField = 'Host';
      return false;
    }

    return true;
  }

  formatValidation() {
    if (!this.isValid()) {
      return {
        message: 'This destination rule has format problems in field ' + this.unformattedField,
        severity: 'error',
        path: ''
      };
    }

    return null;
  }
}
