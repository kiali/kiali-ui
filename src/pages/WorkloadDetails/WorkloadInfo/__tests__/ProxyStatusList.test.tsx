import * as React from 'react';
import { shallow } from 'enzyme';
import ProxyStatusList from '../ProxyStatusList';
import { ProxyStatus } from '../../../../types/Health';
import { Stack, StackItem } from '@patternfly/react-core';
import { shallowToJson } from 'enzyme-to-json';

const syncedProxyStatus: ProxyStatus = {
  CDS: 'Synced',
  EDS: 'Synced',
  LDS: 'Synced',
  RDS: 'Synced'
};

const shallowComponent = (statuses: ProxyStatus) => {
  return shallow(<ProxyStatusList status={statuses} />);
};

describe('ProxyStatusList', () => {
  describe('when status is synced', () => {
    const subject = shallowComponent(syncedProxyStatus);

    it('does not render the stack', () => {
      expect(subject.find(Stack)).toHaveLength(0);
    });

    it('match the snapshot', () => {
      expect(shallowToJson(subject)).toMatchSnapshot();
    });
  });

  describe('when there are unsyced components', () => {
    const statuses: ProxyStatus = syncedProxyStatus;
    syncedProxyStatus.RDS = 'STALE';
    syncedProxyStatus.CDS = 'NOT_SENT';

    const subject = shallowComponent(statuses);

    it('match the snapshot', () => {
      expect(shallowToJson(subject)).toMatchSnapshot();
    });

    it('renders the stack', () => {
      const stack = subject.find(Stack);
      expect(stack).toHaveLength(1);
    });

    it('renders all unsynced statuses', () => {
      const stackItems = subject.find(StackItem);
      expect(stackItems).toHaveLength(3);
      expect(stackItems.at(0).html()).toContain('Istio Proxy Status');
      expect(stackItems.at(1).html()).toContain('CDS: NOT_SENT');
      expect(stackItems.at(2).html()).toContain('RDS: STALE');
    });
  });

  describe('when there are components without value', () => {
    const statuses: ProxyStatus = syncedProxyStatus;
    syncedProxyStatus.RDS = '';
    syncedProxyStatus.CDS = '';

    const subject = shallowComponent(statuses);

    it('match the snapshot', () => {
      expect(shallowToJson(subject)).toMatchSnapshot();
    });

    it('renders all unsynced statuses', () => {
      const stackItems = subject.find(StackItem);
      expect(stackItems).toHaveLength(3);
      expect(stackItems.at(0).html()).toContain('Istio Proxy Status');
      expect(stackItems.at(1).html()).toContain('CDS: -');
      expect(stackItems.at(2).html()).toContain('RDS: -');
    });
  });
});
