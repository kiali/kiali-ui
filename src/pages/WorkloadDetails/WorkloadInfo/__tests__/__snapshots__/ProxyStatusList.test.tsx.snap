// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProxyStatusList when status is synced match the snapshot 1`] = `""`;

exports[`ProxyStatusList when there are components without value match the snapshot 1`] = `
<Stack>
  <StackItem
    className="f9bjrk9"
  >
    Istio Proxy Status
  </StackItem>
  <StackItem
    className="f1e63p1b"
    key="proxy-status-0"
  >
    CDS: -
  </StackItem>
  <StackItem
    className="f1e63p1b"
    key="proxy-status-3"
  >
    RDS: -
  </StackItem>
</Stack>
`;

exports[`ProxyStatusList when there are unsyced components match the snapshot 1`] = `
<Stack>
  <StackItem
    className="f9bjrk9"
  >
    Istio Proxy Status
  </StackItem>
  <StackItem
    className="f1e63p1b"
    key="proxy-status-0"
  >
    CDS: NOT_SENT
  </StackItem>
  <StackItem
    className="f1e63p1b"
    key="proxy-status-3"
  >
    RDS: STALE
  </StackItem>
</Stack>
`;
